Userland/Libraries/LibCompress/LZWDecoder.h

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2018-2021, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2022, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/BitStream.h>
+#include <AK/Debug.h>
+#include <AK/Format.h>
+#include <AK/IntegralMath.h>
+#include <AK/Vector.h>
+
+namespace Compress {
+
+class LZWDecoder {
+private:
+    static constexpr int max_code_size = 12;
+
+public:
+    explicit LZWDecoder(MaybeOwned<LittleEndianInputBitStream> lzw_stream, u8 min_code_size)
+        : m_bit_stream(move(lzw_stream))
+        , m_code_size(min_code_size)
+        , m_original_code_size(min_code_size)
+        , m_table_capacity(AK::exp2<u32>(min_code_size))
+    {
+        init_code_table();
+    }
+
+    u16 add_control_code()
+    {
+        u16 const control_code = m_code_table.size();
+        m_code_table.append(Vector<u8> {});
+        m_original_code_table.append(Vector<u8> {});
+        if (m_code_table.size() >= m_table_capacity && m_code_size < max_code_size) {
+            ++m_code_size;
+            ++m_original_code_size;
+            m_table_capacity *= 2;
+        }
+        return control_code;
+    }
+
+    void reset()
+    {
+        m_code_table.clear();
+        m_code_table.extend(m_original_code_table);
+        m_code_size = m_original_code_size;
+        m_table_capacity = AK::exp2<u32>(m_code_size);
+        m_output.clear();
+    }
+
+    ErrorOr<u16> next_code()
+    {
+        m_current_code = TRY(m_bit_stream->read_bits<u16>(m_code_size));
+
+        if (m_current_code > m_code_table.size()) {
+            dbgln_if(GIF_DEBUG, "Corrupted LZW stream, invalid code: {}, code table size: {}",
```

Comment:
I suppose `GIF_DEBUG` is a bit awkward here now. Maybe we should just remove the log lines for now, or something.

Reply:
I need to touch this file in my next PR, so I will take care of it